### ISSUE TITLE
is("foo"<3) gives internal error

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2297]
Imported from SourceForge on 2024-07-05 23:34:52
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-03-14 16:03:33
Original: https://sourceforge.net/p/maxima/bugs/4120

---

``is`` involving strings gives a Lisp error. It should give a clean Maxima error ... unless of course we start supporting string comparison with ``is``. In that case, what is ``is(3<"foo")``?

We currently have ``is(%i<0)`` => ``false``. That is, if the comparison is meaningless, we return false, because it is not true that %i<0. I would rather see an error, but I know that is controversial.

~~~~
Maxima 5.46.0 https://maxima.sourceforge.io
using Lisp SBCL 2.3.0
Distributed under the GNU Public License. See the file COPYING.
Dedicated to the memory of William Schelter.
The function bug_report() provides bug reporting information.
(%i1) is(3<"foo");

Maxima encountered a Lisp error:

 The value
   "foo"
 is not of type
   LIST

Automatically continuing.
To enable the Lisp debugger set *debugger-hook* to nil.
...
is("foo"<"bar");
<<same>>
~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 23:34:53
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-03-14 18:06:41
Original: https://sourceforge.net/p/maxima/bugs/4120/#c232

---

To fix the Lisp error, MGRP and MGQP (src/compar.lisp) should have special cases for strings -- pretty easy fix. 

MGRP and MGQP look at the sign of the difference of the left and right sides, so I guess anything other than a symbol or cons is going to trigger a similar error -- Lisp arrays for example. Also I wonder if `is` should distribute over lists or anything else. Anyway I suppose at this point I'm going off topic.

As to whether comparing incomparables should  be an error, how about if it's covered by `prederror`. When `prederror` is `true`, an attempt to compare incomparables triggers an error, otherwise return `false`.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 23:34:57
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-04-03 01:02:06
Original: https://sourceforge.net/p/maxima/bugs/4120/#a2f5

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 23:35:01
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-04-03 01:02:06
Original: https://sourceforge.net/p/maxima/bugs/4120/#4b87

---

Fixed by commit [ 61b2abd ]. Closing this report.
